File: pages/settings_page.py
import time
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
from selenium.webdriver.support.ui import WebDriverWait


class SettingsPage:

    _GEAR_BTN = (By.CSS_SELECTOR, 'button:has(svg[data-testid="gearIcon"])')
    _SETTINGS_MENU_ITEM = (By.CSS_SELECTOR, 'a[role="menuitem"]:has(svg[data-testid="gearIcon"])')
    _POPUP_CLOSE_BTN = (By.CSS_SELECTOR, 'button:has(svg[data-icon="xmark-large"])')
    _ADMIN_URL = "/ai-helpy-chat/admin"

    def __init__(self, driver, wait):
        self.driver = driver
        self.wait = wait

    def close_popup_if_present(self):
        try:
            short_wait = WebDriverWait(self.driver, 3)
            popup_text = short_wait.until(
                EC.presence_of_element_located(
                    (By.XPATH, '//*[contains(text(),"개인 토큰 한도가 설정되었습니다.")]')
                )
            )
            assert "개인 토큰 한도가 설정되었습니다." in popup_text.text, "팝업 텍스트 불일치"
            close_btn = short_wait.until(
                EC.element_to_be_clickable(self._POPUP_CLOSE_BTN)
            )
            close_btn.click()
            time.sleep(0.5)
            print("팝업 닫기 완료")
        except TimeoutException:
            pass

    def navigate_to_settings(self):
        self.driver.find_element(*self._GEAR_BTN).click()
        time.sleep(1)

        self.wait.until(EC.element_to_be_clickable(self._SETTINGS_MENU_ITEM)).click()
        time.sleep(1)

        self.wait.until(EC.url_contains(self._ADMIN_URL))
        assert self._ADMIN_URL in self.driver.current_url, \
            f"설정 페이지 이동 실패: {self.driver.current_url}"
        print("설정 페이지 이동 성공:", self.driver.current_url)
        time.sleep(2)

        self.close_popup_if_present()

File: pages/SettingsPage.ts
import { Page, expect } from '@playwright/test'
import { BasePage } from './BasePage'

export class SettingsPage extends BasePage {
  constructor(page: Page) { super(page) }

  get gearButton()       { return this.page.locator('button:has(svg[data-testid="gearIcon"])') }
  get settingsMenuItem() { return this.page.locator('a[role="menuitem"]:has(svg[data-testid="gearIcon"])') }
  get modelTab()         { return this.page.getByRole('tab', { name: /모델 설정|Model/ }) }
  get subscriptionTab()  { return this.page.getByRole('tab', { name: /구독|Subscription/ }) }
  get memberTab()        { return this.page.getByRole('tab', { name: /구성원|Member/ }) }
  get toastMsg()         { return this.page.locator("[class*='toast'], [class*='Snackbar'], [role='alert']").first() }
  get modelToggles()     { return this.page.locator("input[type='checkbox'][name*='model'], input[type='checkbox'][class*='model']") }
  get tokenLimitToggle() { return this.page.locator("input[type='checkbox'][name*='token'], input[type='checkbox'][name*='limit']").first() }

  async navigateToSettings(): Promise<void> {
    await this.gearButton.click()
    await this.settingsMenuItem.click()
    await expect(this.page).toHaveURL(/admin/, { timeout: 10000 })
  }

  async navigateToModelsTab(): Promise<void> {
    await this.modelTab.click()
    await expect(this.page).toHaveURL(/admin\/model/, { timeout: 5000 }).catch(() => {})
  }

  async navigateToSubscriptionTab(): Promise<void> {
    await this.subscriptionTab.click()
  }

  async navigateToMemberTab(): Promise<void> {
    await this.memberTab.click()
  }

  async getToastMessage(): Promise<string> {
    await this.toastMsg.waitFor({ timeout: 5000 })
    return (await this.toastMsg.textContent()) || ''
  }

  async activateDisabledModel(): Promise<string | null> {
    const checkboxes = this.page.locator("input[type='checkbox']")
    const count = await checkboxes.count()
    for (let i = 0; i < count; i++) {
      const cb = checkboxes.nth(i)
      if (!(await cb.isChecked())) {
        await cb.click()
        return `model-${i}`
      }
    }
    return null
  }

  async deactivateActiveModel(): Promise<string | null> {
    const checkboxes = this.page.locator("input[type='checkbox']")
    const count = await checkboxes.count()
    for (let i = 0; i < count; i++) {
      const cb = checkboxes.nth(i)
      if (await cb.isChecked()) {
        await cb.click()
        return `model-${i}`
      }
    }
    return null
  }

  async isSettingsPageDisplayed(): Promise<void> {
    await expect(this.page).toHaveURL(/admin/, { timeout: 10000 })
  }

  async isHistoryPageDisplayed(): Promise<void> {
    await expect(this.page).toHaveURL(/admin\/history/, { timeout: 10000 })
  }

  async navigateToHistoryTab(): Promise<void> {
    await this.page.goto('/ai-helpy-chat/admin/history')
    await this.isHistoryPageDisplayed()
  }
}
